code/object/objectdock.cpp

```diff
@@ -337,6 +339,107 @@ float dock_calc_docked_speed(object *objp)
 	return dfi.maintained_variables.float_value;
 }
 
+void dock_calc_total_moi(matrix* dest, object* objp, vec3d *center)
+{
+	Assertion(dest != nullptr, "dock_calc_total_moi invalid dest");
+	Assertion(objp != nullptr, "dock_calc_total_moi invalid objp");
+
+	*dest = vmd_zero_matrix;
+
+	dock_function_info dfi;
+	dfi.parameter_variables.vecp_value = center;
+	dfi.maintained_variables.matrix_value = dest;
+
+	dock_evaluate_all_docked_objects(objp, &dfi, dock_calc_total_moi_helper);
+}
+
+// This ship is the only ship NOT moved by docking AI to keep everyone together
+// All the other ships in the tree will update based on this one
+// Since this is based on current speed don't expect it to remain consistent between frames
+object* dock_find_dock_root(object *objp)
+{
+	Assertion(objp != nullptr, "dock_find_dock_root invalid argument");
+
+	dock_function_info dfi;
+	object* fastest_objp;
+
+	dfi.maintained_variables.objp_value = nullptr;
+
+	// find the object with the highest speed
+	dock_evaluate_all_docked_objects(objp, &dfi, dock_find_max_speed_helper);
+	fastest_objp = dfi.maintained_variables.objp_value;
+
+	// if we have no max speed, just use the given one
+	if (fastest_objp == nullptr)
+		fastest_objp = objp;
+
+	return fastest_objp;
+}
+
+void dock_calculate_whack_docked_object(vec3d* force, const vec3d* rel_world_hit_pos, object* objp)
+{
+	Assertion((objp != nullptr) && (force != nullptr) && (rel_world_hit_pos != nullptr),
+		"dock_whack_docked_object invalid argument(s)");
+
+	//	Detect null vector.
+	if (whack_below_limit(force))
+		return;
+
+	vec3d world_center_of_mass;
+	vec3d world_hit_pos;
+
+	// calc world hit pos of the hit ship
+	vm_vec_add(&world_hit_pos, rel_world_hit_pos, &objp->pos);
+
+	// calc overall world center-of-mass of all ships
+	float total_mass = dock_calc_docked_center_of_mass(&world_center_of_mass, objp);
+
+	// the new hitpos is the vector from world center-of-mass to world hitpos
+	vm_vec_sub2(&world_hit_pos, &world_center_of_mass);
```

Comment:
Call this new result `world_hit_pos_diff` or even just `hit_pos` because it's not absolutely world coordinates.

Reply:
No, `rel_world_hit_pos` should keep its original name.  I meant that `world_hit_pos` should be renamed.  That's the variable holding the result of the `vm_vec_sub2` operation on this line.